lib/internal/fs/watch/linux.js

```diff
@@ -0,0 +1,244 @@
+
+    try {
+      const files = await opendir(folder);
+
+      for await (const file of files) {
+        const f = path.join(folder, file.name);
+
+        if (this.#closed) {
+          break;
+        }
+
+        if (!this.#files.has(f)) {
+          this.#files.set(f, file);
+          this.emit('change', 'rename', this.#getPath(f));
+
+          if (file.isDirectory()) {
+            await this.#watchFolder(f);
+          } else {
+            this.#watchFile(f);
+          }
+        }
+      }
+    } catch (error) {
+      this.emit('error', error);
+    }
+  }
+
+  #watchFile(file) {
+    validateString(file, 'file');
+
+    const { watchFile } = lazyLoadFsSync();
+
+    if (this.#closed) {
+      return;
+    }
+
+    const existingStat = this.#files.get(file);
+
+    watchFile(file, {
+      persistent: this.#options.persistent,
+    }, (statWatcher, previousStatWatcher) => {
+      if (existingStat && !existingStat.isDirectory() &&
+        statWatcher.nlink !== 0 && existingStat.mtime?.getTime() === statWatcher.mtime?.getTime()) {
+        return;
+      }
+
+      this.#files.set(file, statWatcher);
+
+      if (statWatcher.birthtimeMs === 0 && previousStatWatcher.birthtimeMs !== 0) {
+        // The file is now deleted
+        this.#files.delete(file);
+        this.emit('change', 'rename', this.#getPath(file));
+
+        if (statWatcher.isDirectory()) {
+          this.#unwatchFolder(file);
+        }
+      } else if (statWatcher.isDirectory()) {
+        this.emit('change', 'change', this.#getPath(file));
+        this.#watchFolder(file);
+      } else {
+        this.emit('change', 'change', this.#getPath(file));
+      }
+    });
+  }
+
+  [kFSWatchStart](filename) {
+    validateString(filename, 'filename');
+
+    if (filename.length === 0) {
+      throw new ERR_INVALID_ARG_TYPE('filename');
+    }
+
+    const file = lazyLoadFsSync().statSync(filename);
+
+    this.#rootPath = filename;
+    this.#closed = false;
+    this.#files.set(filename, file);
+
+    if (file.isDirectory()) {
+      traverse(filename, this.#files)
+        .then(() => {
+          for (const f of this.#files.keys()) {
+            this.#watchFile(f);
+          }
+        });
+    } else {
+      this.#watchFile(filename);
+    }
+  }
+
+  ref() {
+    // This is kept to have the same API with FSWatcher
+    throw new ERR_FEATURE_UNAVAILABLE_ON_PLATFORM('ref');
+  }
+
+  unref() {
+    // This is kept to have the same API with FSWatcher
+    throw new ERR_FEATURE_UNAVAILABLE_ON_PLATFORM('unref');
+  }
+
+  async next() {
+    if (this.#closed) {
+      return { __proto__: null, done: true };
+    }
+
+    const result = await new Promise((resolve) => {
+      this.once('change', (eventType, filename) => {
+        resolve({ eventType, filename });
+      });
+    });
+
+    return { __proto__: null, done: false, value: result };
+  }
+
+  // eslint-disable-next-line require-yield
+  *[SymbolAsyncIterator]() {
```

Comment:
```suggestion
  [SymbolAsyncIterator]() {
```

Reply:
This does not work. Changing this line breaks the promise tests, and I could not make it work.

Reply:
What is the failure that you see?

Reply:
This is resolved with the latest changes.